witness/golang/omniwitness/usbarmory/internal/storage/persistance.go

```diff
@@ -0,0 +1,293 @@
+	}
+	if err := p.mapSlot.CheckAndWrite(p.mapWriteToken, smRaw); err != nil {
+		return fmt.Errorf("failed to store mapping: %v", err)
+	}
+	// TODO(al): CheckAndWrite should return the next token rather than us knowing
+	// how the token changes after a successful write.
+	p.mapWriteToken++
+	return nil
+}
+
+// addLog assigns a slot to a new log ID.
+// Must be called with p.mu locked.
+func (p *SlotPersistence) addLog(id string) (uint, error) {
+	if idx, ok := p.idToSlot[id]; ok {
+		return idx, nil
+	}
+	if len(p.freeSlots) == 0 {
+		return 0, errors.New("no free slot available")
+	}
+	f := p.freeSlots[0]
+	p.freeSlots = p.freeSlots[1:]
+	p.idToSlot[id] = f
+	return f, p.storeMap()
+}
+
+// Init sets up the persistence layer. This should be idempotent,
+// and will be called once per process startup.
+func (p *SlotPersistence) Init() error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	s, err := p.part.Open(mappingConfigSlot)
+	if err != nil {
+		return fmt.Errorf("failed to open mapping slot: %v", err)
+	}
+	p.mapSlot = s
+	if err := p.populateMap(); err != nil {
+		return fmt.Errorf("failed to populate logID → slot map: %v", err)
+	}
+	return nil
+}
+
+// Logs returns the IDs of all logs that have checkpoints that can
+// be read.
+func (p *SlotPersistence) Logs() ([]string, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	r := make([]string, 0, len(p.idToSlot))
+	for k := range p.idToSlot {
+		r = append(r, k)
+	}
+	return r, nil
+}
+
+// ReadOps returns read-only operations for the given log ID. This
+// method only makes sense for IDs returned by Logs().
+func (p *SlotPersistence) ReadOps(logID string) (persistence.LogStateReadOps, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	i, ok := p.idToSlot[logID]
+	if !ok {
+		// TODO(al): work around undocumented assumptions in the storage interface;
+		// we have to return a ReadOp even if we don't know about the specified logID.
+		// the ReadOps we return must then return an error of codes.NotFound when GetLatest
+		// is called so that the witness will eventually call WriteOps/Update with the
+		// same logID and create the record.
+		return &slotOps{}, nil
+	}
+	s, err := p.part.Open(i)
+	if err != nil {
+		return nil, fmt.Errorf("internal error opening slot %d associated with log ID %q: %v", i, logID, err)
+	}
+	return &slotOps{slot: s}, nil
+}
+
+// WriteOps shows intent to write data for the given logID. The
+// returned operations must have Close() called when the intent
+// is complete.
+// There is no requirement that the ID is present in Logs(); if
+// the ID is not there and this operation succeeds in committing
+// a checkpoint, then Logs() will return the new ID afterwards.
+func (p *SlotPersistence) WriteOps(logID string) (persistence.LogStateWriteOps, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	i, ok := p.idToSlot[logID]
+	if !ok {
+		var err error
+		i, err = p.addLog(logID)
+		if err != nil {
+			glog.V(2).Infof("Failed to add mapping: %v", err)
```

Comment:
This seems like it should be more like Warningf or Errorf

Reply:
Unfortunately `V(n)` _only_ lets you log at `Info`, and really these verbose log lines are just a work around for higher layers not logging errors.

I'm happy to delete them entirely and we can try to figure out why the errors aren't being logged higher up?

Reply:
Sure, I was actually proposing remove the `V(n)` and logging at those root levels. I don't think anyone running this and getting this error would want to hide it because it was considered too verbose to make the logs readable?

Reply:
That said, we should also work out why this didn't appear in the logs. It may be an artifact of the usbarmory debug tool or how the flags are set or something. From what I can see everything wraps and chucks the error, and then `feeder.go #Run` does a `glog.Errorf` with the result from `FeedOnce`, which I think should be the top level of this stack.

Reply:
ok - I've made them root level `Warning` for now, can ding them when we've figured out the issue with logging on the armory